File: tools/Image.cs
//using System;
//using System.IO;
//using System.Runtime.InteropServices;

//namespace Tools2
//{
//    internal struct RGBColor
//    {
//        public byte B;
//        public byte G;
//        public byte R;
//        public byte A;

//        public RGBColor Diff(RGBColor color)
//        {
//            RGBColor result = new RGBColor();
//            result.A = 255;
//            result.R = (byte)Math.Abs((int)R - (int)color.R);
//            result.G = (byte)Math.Abs((int)G - (int)color.G);
//            result.B = (byte)Math.Abs((int)B - (int)color.B);
//            return result;
//        }

//        public RGBColor Scale(RGBColor color)
//        {
//            RGBColor result = new RGBColor();
//            result.R = Scale(this.R, color.R);
//            result.G = Scale(this.R, color.R);
//            result.B = Scale(this.R, color.R);
//            result.A = this.A;
//            return result;
//        }

//        private static byte Scale(byte value, byte max)
//        {
//            if (max > 0)
//            {
//                float v = (float)value;
//                float m = (float)max;
//                return (byte)(v / m * 255);
//            }
//            return value;
//        }
//    }

//    internal class ImageDiffResult
//    {
//        public string Message { get; set; }
//        public bool IsSame { get; set; }
//        public Image Image { get; set; }
//    }

//    internal class Image
//    {
//        private RGBColor[] data;

//        public ImageDiffResult Diff(Image image)
//        {
//            ImageDiffResult diffResult = new ImageDiffResult();

//            if (this.Width != image.Width || this.Height != image.Height)
//            {
//                diffResult.Message = "Image size does not match";
//                return diffResult;
//            }

//            diffResult.Image = new Image(Width, Height);
//            diffResult.IsSame = true;

//            RGBColor maxColor = new();

//            for (int pos = 0; pos < data.Length; pos++)
//            {
//                RGBColor a = this[pos];
//                RGBColor b = image[pos];
//                var diff = a.Diff(b);
//                diffResult.Image[pos] = diff;

//                if ((diff.R + diff.G + diff.B) > 0)
//                {
//                    diffResult.IsSame = false;
//                }

//                maxColor.R = Math.Max(diff.R, maxColor.R);
//                maxColor.G = Math.Max(diff.G, maxColor.G);
//                maxColor.B = Math.Max(diff.B, maxColor.B);
//            }

//            for (int pos = 0; pos < data.Length; pos++)
//            {
//                data[pos] = data[pos].Scale(maxColor);
//            }

//            return diffResult;
//        }

//        public int Width { get; private set; }
//        public int Height { get; private set; }

//        [StructLayout(LayoutKind.Sequential)]
//        private struct BITMAPINFOHEADER
//        {
//            public uint biSize;
//            public int biWidth;
//            public int biHeight;
//            public ushort biPlanes;
//            public ushort biBitCount;
//            public uint biCompression;
//            public uint biSizeImage;
//            public int biXPelsPerMeter;
//            public int biYPelsPerMeter;
//            public uint biClrUsed;
//            public uint biClrImportant;
//        }

//        [StructLayout(LayoutKind.Sequential, Pack = 2)]
//        private struct BITMAPFILEHEADER
//        {
//            public ushort bfType;
//            public uint bfSize;
//            public ushort bfReserved1;
//            public ushort bfReserved2;
//            public uint bfOffBits;
//        }

//        public Image(int width, int height)
//        {
//            this.Width = width;
//            this.Height = height;
//            this.data = new RGBColor[(width * height)];
//        }

//        public Image(string fileName)
//        {
//            using var reader = new BinaryReader(File.Open(fileName, FileMode.Open));

//            var fh = reader.ReadBytes(Marshal.SizeOf(typeof(BITMAPFILEHEADER)));
//            var bh = reader.ReadBytes(Marshal.SizeOf(typeof(BITMAPINFOHEADER)));

//            var bfh = GetStruct<BITMAPFILEHEADER>(fh);
//            var bih = GetStruct<BITMAPINFOHEADER>(bh);

//            this.Width = Math.Abs(bih.biWidth);
//            this.Height = Math.Abs(bih.biHeight);

//            byte[] byteData = reader.ReadBytes(Width * Height * 4);

//            this.data = new RGBColor[(Width * Height)];

//            for (int y = 0; y < Height; y++)
//            {
//                for (int x = 0; x < Width; x++)
//                {
//                    var pos = (x + Width * y);
//                    this.data[pos] = new RGBColor
//                    {
//                        B = byteData[4 * pos + 0],
//                        G = byteData[4 * pos + 1],
//                        R = byteData[4 * pos + 2],
//                        A = byteData[4 * pos + 3]
//                    };
//                }
//            }
//        }

//        public RGBColor this[int index]
//        {
//            get { return data[index]; }
//            set { data[index] = value; }
//        }

//        public void Save(string fileName)
//        {
//            var infoHeaderSize = Marshal.SizeOf(typeof(BITMAPINFOHEADER));
//            var fileHeaderSize = Marshal.SizeOf(typeof(BITMAPFILEHEADER));
//            var offBits = infoHeaderSize + fileHeaderSize;

//            BITMAPINFOHEADER infoHeader = new BITMAPINFOHEADER
//            {
//                biSize = (uint)infoHeaderSize,
//                biBitCount = 32,
//                biClrImportant = 0,
//                biClrUsed = 0,
//                biCompression = 0,
//                biHeight = -Height,
//                biWidth = Width,
//                biPlanes = 1,
//                biSizeImage = (uint)(Width * Height * 4)
//            };

//            BITMAPFILEHEADER fileHeader = new BITMAPFILEHEADER
//            {
//                bfType = 'B' + ('M' << 8),
//                bfOffBits = (uint)offBits,
//                bfSize = (uint)(offBits + infoHeader.biSizeImage)
//            };

//            using var writer = new BinaryWriter(File.Open(fileName, FileMode.Create));
//            writer.Write(GetBytes(fileHeader));
//            writer.Write(GetBytes(infoHeader));
//            foreach (var color in data)
//            {
//                writer.Write(color.B);
//                writer.Write(color.G);
//                writer.Write(color.R);
//                writer.Write(color.A);
//            }
//        }

//        private static byte[] GetBytes<T>(T data)
//        {
//            var length = Marshal.SizeOf(data);
//            var ptr = Marshal.AllocHGlobal(length);
//            var result = new byte[length];
//            Marshal.StructureToPtr(data, ptr, true);
//            Marshal.Copy(ptr, result, 0, length);
//            Marshal.FreeHGlobal(ptr);
//            return result;
//        }

//        private T GetStruct<T>(byte[] bytes) where T : struct
//        {
//            T stuff;
//            GCHandle handle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
//            try
//            {
//                stuff = (T)Marshal.PtrToStructure(handle.AddrOfPinnedObject(), typeof(T));
//            }
//            finally
//            {
//                handle.Free();
//            }
//            return stuff;
//        }
//    }
//}

File: ruby/Image.rb
class BinaryStream
    def initialize()
    end

    def write_word(value)
    end

    def write_dword(value)
    end

    def write_long(value)
    end
end

class BitmapInfoHeader
    def initialize(w, h)
        @biSize = 0          # DWORD
        @biWidth = 0         # LONG
        @biHeight = 0        # LONG
        @biPlanes = 0        # WORD
        @biBitCount = 0      # WORD
        @biCompression = 0   # DWORD
        @biSizeImage = 0     # DWORD
        @biXPelsPerMeter = 0 # LONG
        @biYPelsPerMeter = 0 # LONG
        @biClrUsed = 0       # DWORD
        @biClrImportant = 0  # DWORD

        @biSize = 40;
        @biWidth = w;
        @biHeight = -h;
        @biPlanes = 1;
        @biBitCount = 32;
        @biSizeImage = w * h * 4;
    end

    def write_to (stream)
        stream.write_dword(@biSize),
        stream.write_long(@biWidth),
        stream.write_long(@biHeight),
        stream.write_word(@biPlanes),
        stream.write_word(@biBitCount),
        stream.write_dword(@biCompression),
        stream.write_dword(@biSizeImage),
        stream.write_long(@biXPelsPerMeter),
        stream.write_long(@biYPelsPerMeter),
        stream.write_dword(@biClrUsed),
        stream.write_dword(@biClrImportant),
    end
end

class BitmapFileHeader
    # private bfType: number = 0;    // WORD
    # public bfSize: number = 0;     // DWORD
    # public bfReserved: number = 0; // DWORD
    # public bfOffBits: number = 0;  // DWORD

    def initialize(imageSize: number)
        @bfType = 0x4D42;
        @bfOffBits = 54;
        @bfSize = @bfOffBits + imageSize;
    end

    def write_to(stream)
        stream.write_word(@bfType),
        stream.write_dword(@bfSize),
        stream.write_dword(@bfReserved),
        stream.write_dword(@bfOffBits),
    end
end


class Image
    def initialize(width, height)
        @width = width
        @height = height
        @colors = Array.new(width) { Array.new(height) }
    end

    def set_pixel(x, y, color)
        @colors[x][y] = color
    end

    def save_to(filname)
        data = Int8Array.new(54 + @data.width * @data.height * 4);
        stream = BinaryStream.new(data)

        bih = BitmapInfoHeader.new(this.width, this.height);
        bfh = BitmapFileHeader.new(this.data.length);

        bfh.write_to(stream)
        bih.write_to(stream)

        # write bitmap bites

        # save image
    end
end